mlkem/poly.c

```diff
@@ -274,28 +274,38 @@ void poly_frombytes(poly *r, const uint8_t a[MLKEM_POLYBYTES]) {
 }
 #endif /* MLKEM_USE_NATIVE_POLY_FROMBYTES */
 
-/*************************************************
- * Name:        poly_frommsg
- *
- * Description: Convert 32-byte message to polynomial
- *
- * Arguments:   - poly *r: pointer to output polynomial
- *              - const uint8_t *msg: pointer to input message
- **************************************************/
 void poly_frommsg(poly *r, const uint8_t msg[MLKEM_INDCPA_MSGBYTES]) {
-  unsigned int i, j;
-
 #if (MLKEM_INDCPA_MSGBYTES != MLKEM_N / 8)
 #error "MLKEM_INDCPA_MSGBYTES must be equal to MLKEM_N/8 bytes!"
 #endif
 
-  for (i = 0; i < MLKEM_N / 8; i++) {
-    for (j = 0; j < 8; j++) {
-      r->coeffs[8 * i + j] = 0;
-      cmov_int16(r->coeffs + 8 * i + j, ((MLKEM_Q + 1) / 2), (msg[i] >> j) & 1);
-    }
-  }
+  const int num_blocks = MLKEM_N / 8;
+
+  for (int i = 0; i < num_blocks; i++)
+      // clang-format off
+    ASSIGNS(i, OBJECT_WHOLE(r))
+    INVARIANT(i >= 0)
+    INVARIANT(i <= num_blocks)
+    INVARIANT(num_blocks == 32)
+    INVARIANT(ARRAY_IN_BOUNDS(int, k, 0, (8 * i - 1), r->coeffs, 0, (MLKEM_Q - 1)))
+    // clang-format on
+    {
+      for (int j = 0; j < 8; j++)
+          // clang-format off
+        ASSIGNS(j, OBJECT_WHOLE(r))
+        INVARIANT(i >= 0 && i <  num_blocks)
+        INVARIANT(j >= 0 && j <= 8)
+        INVARIANT(ARRAY_IN_BOUNDS(int, k, 0, (8 * i + j - 1), r->coeffs, 0, (MLKEM_Q - 1)))
+        // clang-format on
+        {
+          r->coeffs[8 * i + j] = 0;
+          cmov_int16(&r->coeffs[8 * i + j], HALF_Q, (msg[i] >> j) & 1);
 
+          // Check that the value just assigned must be either HALF_Q or 0
```

Comment:
This can be removed? It is a verbatim restatement of the post condition of `cmov_int16`?

Reply:
Yes... removing it does not seem to change proof performance, so it can go.

Reply:
I took the liberty to change that.

Reply:
OK